### ISSUE TITLE
[DX] hook_field_*_settings_form() should have $form and $form_state parameters

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#3590]
This is the respective issue as https://www.drupal.org/node/1391196

> The hooks for instance and widget settings, hook_field_widget_settings_form() and hook_field_instance_settings_form() both return form arrays, but receive no information at all about the larger form they're inserting into.
>
>This means that they can't add more complex behaviours such as dependencies and ajax -- hook_form_alter has to be used instead.

That ticket was switched from D8 to D7 which back in 2013, with this comment:

> Field types are no plugins, so this doesn't apply anymore to D8. I'm not even sure whether we can change those function signatures in D7 though ... feels like a won't fix to me.

Well, thankfully we have no problem with API *additions* in Backdrop (so long as we are not breaking backwards compatibility)...

---
PR by @klonos: https://github.com/backdrop/backdrop/pull/2533

**Comment by klonos** [comment]
So how this whole thing started was that I started doing some research and googling around in order to see if #3379 would be something I could work on. That has eventually lead me to https://www.drupal.org/docs/7/api/javascript-api/ajax-forms-in-drupal-7 which seemed like something I could give a go.

So it looked like things would need to be changed in `text_field_instance_settings_form()`, but I quickly realised that there was no (easy) way to get any information about the form we're injecting these field instance settings into. Hence, this issue here 😄

**Comment by herbdool** [comment]
Tests passing and it seems like it's backwards compatible.

**Comment by klonos** [comment]
Thanks @herbdool 👍 ...proposing 1.13.0 for milestone.

**Comment by quicksketch** [comment]
Seems good to me. Could we add a `@since` docblock item here: 

https://github.com/backdrop/backdrop/pull/2533/files#r267928815

**Comment by klonos** [comment]
@quicksketch done! 😄 ...I have added a bit more info in the docblock as per why this change, and a link to this issue here.

PS: The sooner this gets in, the sooner I may be able to take a shot at #3379, which I'd also like to make it into 1.13.0 if possible.

**Comment by jenlampton** [comment]
I agree with the milestone candidate, but do we want to shoot for a minor or bug-fix release?

**Comment by BWPanda** [comment]
PR has conflicts so needs rebasing. Also, was @herbdool's PR comment addressed?
